trajectory_executor/src/trajectory_executor_node.cpp

```diff
@@ -0,0 +1,185 @@
+
+    timer_ = create_timer(
+        get_clock(),
+        std::chrono::milliseconds(timer_period_ms), 
+        std::bind(&TrajectoryExecutor::onTrajEmitTick, this));
+
+    return CallbackReturn::SUCCESS;
+  }
+
+  void TrajectoryExecutor::onTrajEmitTick()
+  {
+    std::unique_lock<std::mutex> lock(cur_traj_mutex_);
+    RCLCPP_DEBUG_STREAM(get_logger(), "TrajectoryExecutor tick start!");
+
+    if (cur_traj_ != nullptr) {
+      if (!cur_traj_->trajectory_points.empty()) {
+        // Determine the relevant control plugin for the current timestep
+        std::string control_plugin = cur_traj_->trajectory_points[0].controller_plugin_name;
+        // if it instructed to use default control_plugin
+        if (control_plugin == "default" || control_plugin =="") {
+          control_plugin = config_.default_control_plugin;
+        }
+
+        std::map<std::string, carma_ros2_utils::PubPtr<carma_planning_msgs::msg::TrajectoryPlan>>::iterator it = traj_publisher_map_.find(control_plugin);
+        if (it != traj_publisher_map_.end()) {
+          RCLCPP_DEBUG_STREAM(get_logger(), "Found match for control plugin " << control_plugin.c_str() << " at point " << timesteps_since_last_traj_ << " in current trajectory!");
+          it->second->publish(*cur_traj_);
+        } else {
+          std::ostringstream description_builder;
+                description_builder << "No match found for control plugin " 
+                << control_plugin << " at point " 
+                << timesteps_since_last_traj_ << " in current trajectory!";
+
+          throw std::invalid_argument(description_builder.str());
+        }
+        timesteps_since_last_traj_++;
+      } else {
+        throw std::out_of_range("Ran out of trajectory data to consume!");
+      }
+    } else {
+      RCLCPP_DEBUG_STREAM(get_logger(), "Awaiting initial trajectory publication...");
+    }
+    RCLCPP_DEBUG_STREAM(get_logger(), "TrajectoryExecutor tick completed succesfully!");
+
+  }
+
+  void TrajectoryExecutor::onNewTrajectoryPlan(carma_planning_msgs::msg::TrajectoryPlan::UniquePtr msg)
+  {
+    std::unique_lock<std::mutex> lock(cur_traj_mutex_); // Acquire lock until end of this function scope
+    RCLCPP_DEBUG_STREAM(get_logger(), "Received new trajectory plan!");
+    RCLCPP_DEBUG_STREAM(get_logger(), "New Trajectory plan ID: " << msg->trajectory_id);
+    RCLCPP_DEBUG_STREAM(get_logger(), "New plan contains " << msg->trajectory_points.size() << " points");
+
+    cur_traj_ = std::unique_ptr<carma_planning_msgs::msg::TrajectoryPlan>(move(msg));
+    timesteps_since_last_traj_ = 0;
+    RCLCPP_INFO_STREAM(get_logger(), "Successfully swapped trajectories!");  
+  }
+
+  void TrajectoryExecutor::guidanceStateMonitor(carma_planning_msgs::msg::GuidanceState::UniquePtr msg)
+  {
+    std::unique_lock<std::mutex> lock(cur_traj_mutex_); // Acquire lock until end of this function scope
```

Comment:
We can move this mutex inside the conditional, that way we only acquire it when we need it. Saves us from doing an unneeded syscall - or blocking in the worst case - when it's not needed.

Reply:
Fixed.